From 0f2d5ef17040bc68e355f177b42b0e44e2cfc843 Mon Sep 17 00:00:00 2001
From: Peter Zijlstra <peterz@infradead.org>
Date: Mon, 3 Oct 2016 16:35:32 +0200
Subject: [PATCH 03/11] sched: Add missing update_rq_clock() call for
 task_hot().
Git-commit: 3bed5e2166a5e433bf62162f3cd3c5174d335934
Patch-mainline: v4.11-rc1
References: bsc#981825

Add the update_rq_clock() call at the top of the callstack instead of
at the bottom where we find it missing, this to aid later effort to
minimize the number of update_rq_lock() calls.

4WARNING: CPU: 30 PID: 194 at ../kernel/sched/sched.h:797 assert_clock_updated.isra.63.part.64+0x23/0x30
rq->clock_update_flags < RQCF_ACT_SKIPdModules linked in:
dCPU: 30 PID: 194 Comm: kworker/30:0 Not tainted 4.8.0-00637-g67223e2-dirty #554
dHardware name: Intel Corporation S2600GZ/S2600GZ, BIOS SE5C600.86B.02.02.0002.122320131210 12/23/2013
 ffffc90003963b90 ffffffff816152e5 ffffc90003963be0 0000000000000000
 ffffc90003963bd0 ffffffff810d5bab 0000031d00002e04 ffffc90003963ce0
 ffff88082f357b00 ffff88082f357b00 ffff88082d570000 0000000000017b00
Call Trace:
 [<ffffffff816152e5>] dump_stack+0x67/0x92
 [<ffffffff810d5bab>] __warn+0xcb/0xf0
 [<ffffffff810d5c1f>] warn_slowpath_fmt+0x4f/0x60
 [<ffffffff81107113>] assert_clock_updated.isra.63.part.64+0x23/0x30
 [<ffffffff811078e7>] can_migrate_task+0x2b7/0x2d0
 [<ffffffff811129fe>] load_balance+0x41e/0x9f0
 [<ffffffff8111327e>] pick_next_task_fair+0x2ae/0x510
 [<ffffffff81a848eb>] __schedule+0xeb/0x820
 [<ffffffff810fbc93>] ? finish_task_switch+0x83/0x220
 [<ffffffff810effd0>] ? process_one_work+0x4a0/0x4a0
 [<ffffffff81a85060>] schedule+0x40/0x90
 [<ffffffff810f008f>] worker_thread+0xbf/0x4e0
 [<ffffffff810effd0>] ? process_one_work+0x4a0/0x4a0
 [<ffffffff810effd0>] ? process_one_work+0x4a0/0x4a0
 [<ffffffff810f5b8a>] kthread+0xca/0xe0
 [<ffffffff81a8961c>] ? _raw_spin_unlock_irq+0x1c/0x30
 [<ffffffff810f5ac0>] ? kthread_park+0x60/0x60
 [<ffffffff81a89bc7>] ret_from_fork+0x27/0x40
4---[ end trace 3a72b08178dc732c ]---

Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
Signed-off-by: Matt Fleming <mfleming@suse.de>
---
 kernel/sched/fair.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index bfa1e44..a1a0b29 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -7432,6 +7432,7 @@ redo:
 
 more_balance:
 		raw_spin_lock_irqsave(&busiest->lock, flags);
+		update_rq_clock(busiest);
 
 		/*
 		 * cur_ld_moved - load moved in current iteration
@@ -7797,6 +7798,7 @@ static int active_load_balance_cpu_stop(void *data)
 		};
 
 		schedstat_inc(sd, alb_count);
+		update_rq_clock(busiest_rq);
 
 		p = detach_one_task(&env);
 		if (p) {
-- 
2.10.0

